File: parameter_calc/new_parameter_calc/parameter_calc/matlab/functions/ceq.m
function F = ceq(X, xb, F_terrain_env, p)
% ceq  역기구학 제약 방정식 (griddedInterpolant 기반 초고속 평가)

    [Wf, Wm, Wr, ~, ~] = wpos(X, xb, p);

    % [수정] F_terrain_env는 Minkowski 팽창이 적용되어 이미 R_w만큼 높아져 있음 (+ p.R_w 제거)
    hf = F_terrain_env(Wf(1));
    hm = F_terrain_env(Wm(1));
    hr = F_terrain_env(Wr(1));

    F = [Wf(2) - hf;
         Wm(2) - hm;
         Wr(2) - hr];
end

File: parameter_calc/new_parameter_calc/parameter_calc/python/functions/ceq.py
"""
ceq: 역기구학 제약 방정식 (interpolant 기반)
"""
import numpy as np
from .wpos import wpos


def ceq(X, xb, F_terrain_env, p):
    Wf, Wm, Wr, _, _ = wpos(X, xb, p)

    hf = F_terrain_env(Wf[0])
    hm = F_terrain_env(Wm[0])
    hr = F_terrain_env(Wr[0])

    return np.array([Wf[1] - hf, Wm[1] - hm, Wr[1] - hr])
